src/python/grpcio/grpc/_runtime_protos.py

```diff
@@ -15,6 +15,48 @@
 import sys
 
 _REQUIRED_SYMBOLS = ("_protos", "_services", "_protos_and_services")
+_MINIMUM_VERSION = (3, 5, 0)
+
+
+def _has_runtime_proto_symbols(mod):
+    return all(hasattr(mod, sym) for sym in _REQUIRED_SYMBOLS)
+
+
+def _is_grpc_tools_importable():
+    try:
+        import grpc_tools  # pylint: disable=unused-import
+        return True
+    except ImportError as e:
+        # NOTE: It's possible that we're encountering a transitive ImportError, so
+        # we check for that and re-raise if so.
+        if "grpc_tools" not in e.args[0]:
+            raise
+        return False
+
+
+def _call_with_lazy_import(fn_name, version_fn, uninstalled_fn, protobuf_path):
+    """Calls one of the three functions, lazily importing grpc_tools.
+
+    Args:
+      fn_name: The name of the function to import from grpc_tools.protoc.
+      version_fn: A function to call in case the Python version is insufficient.
+      uninstalled_fn: A function to call in case grpc_tools is not installed.
+      protobuf_path: The path to import.
+
+    Returns:
+      The appropriate module object.
+    """
+    if sys.version_info < _MINIMUM_VERSION:
+        return version_fn(protobuf_path)
+    else:
+        if not _is_grpc_tools_importable():
+            return uninstalled_fn(protobuf_path)
+        import grpc_tools.protoc
+        if _has_runtime_proto_symbols(grpc_tools.protoc):
+            fn = getattr(grpc_tools.protoc, fn_name)
+            return fn(protobuf_path)
+        else:
+            return uninstalled_fn(protobuf_path)
```

Comment:
nit: What's the reason for passing `protobuf_path` to `*_fn`?

Reply:
No real reason. The argument is ignored, but I left it there for consistency of call pattern with the actual function. Is there some advantage to *not* passing this argument I'm not seeing. All I can think of is a small speed boost, but this is an error case in an import, so that argument doesn't seem particularly compelling to me.

Reply:
If you feel strongly about it, I can remove the argument.

Reply:
No, isn't fine.

It wasn't used, so I thought your intention was to log it, like "Error while generating for file [...]:". Otherwise, we could consider make those exception raising functions taking `unused_args`, `unused_kwargs`.

Reply:
I went ahead and factored out the error functions entirely since they're not really necessary anymore.